### ISSUE TITLE
Error in BatchNorm output during test time

### ISSUE THREAD

**Issue by ayush1999** [FluxML/Flux.jl#314]
During test time, the current batchnorm output is given by : `λ.(reshape(γ, affine_shape...) .* ((x .- μ) ./ σ) .+ reshape(β, affine_shape...))`, where `σ =  reshape(BN.σ, affine_shape...)`.

However, during the actual value of `σ` should be `sqrt.(reshape(BN.σ, affine_shape...) .+ BN.ϵ)`. (Referring to the original BatchNorm paper). 

I need a confirmation on this before I go ahead and open a PR.
cc @phitomson @MikeInnes

**Comment by CarloLucibello** [comment]
The current implementation of batch norm is consistent, since  `BN.σ` is already a running average of the standard deviation (so no need for an extra square root):

```julia
    μ = mean(x, axes)
    σ = sqrt.(mean((x .- μ).^2, axes) .+ ϵ)

    # update moving mean/std
    mtm = data(convert(T, BN.momentum))
    BN.μ = (1 - mtm) .* BN.μ .+ mtm .* squeeze(data(μ), (axes...))
    BN.σ = (1 - mtm) .* BN.σ .+ mtm .* squeeze(data(σ), (axes...)) .* m ./ (m - 1)
```
There is a little problem though: the correction m/(m-1) to the estimator makes sense only for the variance, while for the stddev should be `sqrt(m/(m-1))` instead.
Also, for reproducibilty, I would prefer to adhere to the paper and keep the running average of the variance, and remove that extra epsilon. So I would change the code to
```julia
    μ = mean(x, axes)
    σ² = var(x, axes, mean=μ)

    # update moving mean/var
    mtm = data(convert(T, BN.momentum))
    BN.μ = (1 - mtm) .* BN.μ .+ mtm .* squeeze(data(μ), (axes...))
    BN.σ² = (1 - mtm) .* BN.σ² .+ mtm .* squeeze(data(σ²), (axes...)) .* m ./ (m - 1)
```

**Comment by ayush1999** [comment]
@CarloLucibello I am talking about the test part of the code. That is when the `active` attribute is `False`. What you described was the training time.
During testing time, the flow doesn't go to the part you described (since it is in the `if-else` condition.).
Hence, we never calculate the root of `variance+epsilon`. We merely divide it by `σ`.

**Comment by CarloLucibello** [comment]
Right, I'm saying it is correct to divide by σ at test time since σ is already a standard deviation

**Comment by ayush1999** [comment]
@CarloLucibello And what about `BN.ϵ`? At test time, we should divide by `sqrt(variance + BN.ϵ)`. Why are we ignoring `BN.ϵ` in this case?

**Comment by ayush1999** [comment]
In that case, as you said, @CarloLucibello , we should have 
```
let λ = BN.λ
    λ.(reshape(γ, affine_shape...) .* ((x .- μ) ./ sqrt.(σ.^2 .+ BN.ϵ)) .+ reshape(β, affine_shape...))
  end
```
Instead of  https://github.com/FluxML/Flux.jl/blob/master/src/layers/normalise.jl#L143.

**Comment by CarloLucibello** [comment]
right now we are not using `BN.ϵ` at test time because we add it up to `σ` at training time. As I said in my first comment the current implementation is consistent (except for the m/(m-1)), but I'd prefer to switch to the variance based version

**Comment by MikeInnes** [comment]
@CarloLucibello would you be willing to put up a quick PR? Definitely good to be consistent with CUDNN and other frameworks.
